Also listen on loading images

elemoine opened the pull request:
This fixes a bug that occured when an image source was used by multiple maps. In that case the map that didn't load the image wouldn't register a load listener on that image and would therefore not call render to request a re-render of the map. I also took the opportunity to factor out common code in at function implemented in the layer renderer base class.

As discussed during the dev meeting this patch relies on that fact that `goog.events.listen` would not register a new listener if it already has a similar listener (with the same callback function etc.) See https://github.com/google/closure-library/blob/master/closure/goog/events/listenermap.js#L80-L97.

This fixes a bug reported on the mailing list: <https://groups.google.com/d/msgid/ol3-dev/1686b2f3-a946-4e25-b572-9b93aa38e513%40googlegroups.com?utm_medium=email&utm_source=footer>.

Please review.

tschaub commented on src/ol/renderer/layerrenderer.js:
```diff
@@ -95,6 +97,33 @@ ol.renderer.Layer.prototype.handleImageChange = function(event) {
+  } else {
+    if (imageState == ol.ImageState.IDLE) {
+      image.load();
+    }
```
If the image is cached, does the listener have to be registered before load is called?  There used to be browser issues with `src` getting set before the load handler (load handler wouldn't get called if the image was cached).

elemoine replied:
Right. http://stackoverflow.com/questions/12354865/image-onload-event-and-browser-cache also mentions that problem. I'll change the order: register the listener first, and then call load if the image is idle.

elemoine replied:
But when the same image source is used with multiple maps the maps that won't load the image (because it was loaded by another map) will still register their listener after the image `src` is set.

tschaub commented on src/ol/renderer/layerrenderer.js:
```diff
@@ -95,6 +97,33 @@ ol.renderer.Layer.prototype.handleImageChange = function(event) {
+  if (imageState == ol.ImageState.LOADED ||
+      imageState == ol.ImageState.ERROR) {
+    goog.events.unlisten(image, goog.events.EventType.CHANGE,
+        this.handleImageChange_, false, this);
```
It's a bit odd to see this unregistration in a load method.  This relies on `loadImage` getting called multiple times?

elemoine replied:
For a given image and a given layer renderer we do want the listener to be called just once, so `listenOnce` is indeed what we want to use here.

Just like `listen`, `listenOnce` will not create a new listener when it is called again with the same (target/image, listener, scope/renderer, capture) tuple.

So I'll try to use `listenOnce` again.

elemoine commented:
I updated the patch based on @tschaub's comments (and my responses). I tested the http://jsfiddle.net/38q1q3j0/10/ test-case in Chrome, Firefox, IE 9 and IE 10, and the jsfiddle works as expected in these browsers. And I added unit-tests for `ol.renderer.Layer#loadImage`.

Please review again.

elemoine commented:
@tschaub, do you have more comments?
